pjmedia/src/pjmedia/sdp_neg.c

```diff
@@ -1545,6 +1571,106 @@ static pj_status_t create_answer( pj_pool_t *pool,
+ * 1: pt found
+ * 2: codec found
+ * Return (bitwise):
+ * 0: pt and codec not found
+ * 1: only pt found
+ * 2: only codec found
+ * 3: pt and codec found
+ */
+static unsigned pt_exists(const media_pt_map *sess_med_pt_map, unsigned pt, 
+                          const pj_str_t* codec)
+{
+    unsigned i = 0;
+    unsigned cnt = sess_med_pt_map->med_pt_map_cnt;
+    unsigned max_cnt = PJ_ARRAY_SIZE(sess_med_pt_map->med_pt_map);
+    unsigned retval = 0;
+    for (; i < cnt && i < max_cnt; ++i) {
+        unsigned tmp_ret = 0;
+
+        if (sess_med_pt_map->med_pt_map[i].pt == pt) {
+            tmp_ret |= 1;
+        }
+        if (pj_strcmp(&sess_med_pt_map->med_pt_map[i].codec, codec) == 0) {
+            tmp_ret |= 2;
+        }
+
+        if (tmp_ret == 3)
+            return tmp_ret;
+
+        if (tmp_ret > retval)
+            retval = tmp_ret;
+    }
+    return retval;
+}
+
+#if PJMEDIA_SDP_NEG_MAINTAIN_SESSION_PT
+
+static media_pt_map* get_media_pt_map(pjmedia_sdp_neg* neg, pj_str_t* med_type)
+{
+    unsigned i = 0;
+    unsigned max_cnt = PJ_ARRAY_SIZE(neg->sess_med_pt_map);
+    for (; i < neg->sess_med_pt_map_cnt && i < max_cnt; ++i) {
+        if (pj_strcmp(&neg->sess_med_pt_map[i].med_type, med_type) == 0) {
+            break;
+        }
+    }
+    if (i < max_cnt) {
+        if (i == neg->sess_med_pt_map_cnt) {
+            pj_strdup(neg->pool_neg, &neg->sess_med_pt_map[i].med_type, med_type);
+            neg->sess_med_pt_map_cnt++;
+        }
+        return &neg->sess_med_pt_map[i];
+    }
+    return NULL;
+}
+
+static void update_media_pt_map(const pjmedia_sdp_session* sess,
+                                pjmedia_sdp_neg* neg)
+{
+    unsigned i = 0;
+
+    for (; i < sess->media_count; ++i) {
+        unsigned j = 0;
+        pjmedia_sdp_media* sess_media = sess->media[i];
+        media_pt_map* med_pt_map = get_media_pt_map(neg, &sess->media[i]->desc.media);
+
+        if (!med_pt_map)
+            return;
+
+        for (; j < sess_media->desc.fmt_count; ++j) {
+            /* Get the codec id */
+            const pjmedia_sdp_attr* attr;
+            pj_str_t codec;
+            unsigned pt = pj_strtoul(&sess_media->desc.fmt[j]);
+            unsigned pt_len;
+
+            if (pt < START_DYNAMIC_PT)
+                continue;
+
+            attr = pjmedia_sdp_media_find_attr2(sess_media, "rtpmap",
+                                                &sess_media->desc.fmt[j]);
+
+            if (!attr) {
+                continue;
+            }
+            pt_len = sess_media->desc.fmt[j].slen + 1;
+            codec.ptr = attr->value.ptr + pt_len;
+            codec.slen = attr->value.slen - pt_len;
+
+            if (pt_exists(med_pt_map, pt, &codec) != 3) {
+                med_pt_map->med_pt_map[med_pt_map->med_pt_map_cnt].pt = pt;
+                pj_strdup(neg->pool_neg,
+                          &med_pt_map->med_pt_map[med_pt_map->med_pt_map_cnt++].codec,
+                          &codec);
+            }
```

Comment:
- If there is a reused PT for different codecs, that PT will be mapped to two codecs, is this correct?
- Better put some comments about the behavior for future ref, also better to explain why.

Reply:
Yes, that is correct. This is to handle if remote might send a conflicting PT/codec.

Reply:
Then how we decide for that shared/reused PT, e.g: used for which codec?

Reply:
Currently, pick the first PT assigned to the codec.